core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -2025,7 +2029,17 @@ public List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix
     @Override
     public Collection<FunctionMetadata> listFunctions(Session session)
     {
-        return functions.listFunctions();
+        ImmutableList.Builder<FunctionMetadata> functions = ImmutableList.builder();
+        functions.addAll(this.functions.listFunctions());
+        for (SqlPathElement sqlPathElement : session.getPath().getParsedPath()) {
+            String catalog = sqlPathElement.getCatalog().map(Identifier::getValue).or(session::getCatalog)
+                    .orElseThrow(() -> new IllegalArgumentException("Session default catalog must be set to resolve a partial function name: " + null));
```

Comment:
Throw a TrinoException. Also, the `+ null` at the end of the message seems like a mistake.